python/across_api/auth/api.py

```diff
@@ -33,10 +33,11 @@ def __init__(self, **kwargs):
 
         super().__init__(**kwargs)
 
-    async def __call__(self, request: Request) -> None:
+    async def __call__(self, request: Request):
         credentials: Optional[HTTPAuthorizationCredentials] = await super().__call__(
             request
         )
```

Comment:
add `-> Optional[dict]` type decoration to `__call__`.

Reply:
I changed it to `Optional[dict]` as the type decoration, and mypy is still throwing a fit about it. I believe it is from the  inheritance of `HTTPBearer` and how its `__call__` might have a return type that we aren't meeting in the decorator. I just threw a ` # type: ignore` on it. This should be sufficient since a return type shouldn't be expected how we are implementing our nested dependencies for this case.

Reply:
Can you change it to `# type: ignore[override]` so we're telling `mypy` only to ignore this one specific issue, rather than ignoring all issues that might crop up.

Reply:
OK, sorry I just stared at this code for 5 more minutes. The root cause here is that you're forcing this function to return a dictionary of the access token, whereas it should just pass through the value of `credentials`. If you change the return to Optional[HTTPAuthorizationCredentials] and just return `credentials` at the end, and make no other returns, it works fine. 

Then you can change `ScopeAuthenticate` to start like this:
```
async def ScopeAuthenticate(
    security_scopes: SecurityScopes,
    credentials: Annotated[HTTPAuthorizationCredentials, Depends(JWTBearerSecurity)],
):
    valid = False
    access_token = jwt.get_unverified_claims(credentials.credentials)
    scopes = access_token.get("scope", "")
    ...
```

Reply:
I'm fairly nervous about calling `jwt.get_unverified_claims` in a function that is separate from the one that checks the signature. It makes it less obvious that we have ever checked the signature.

Could you instead have `JWTBearerSecurity` depend upon `HTTPBearer` rather than inheriting from it?

Reply:
I can understand the appearance in insecurity, however `credentials` doesn't get passed to `ScopeAuthenticate` unless its signature verification is passed. If you treat `get_unverified_claim` for what it is, a way of extracting a key from a JWT without doing another signature check, it's fine. If we're going to rework the whole code to allay his fear, I'd just favor leaving as is.

Reply:
I actually don't know whether this is secure or not. We used `jwt.decode` in the other context, not `get_unverified_claims`. I'd have to consult the implementation of `get_unverified_claims` to determine if it only harvests claims from the signed section of the JWT (though it probably does). However this code _is_ insecure in isolation, and it creates the risk that we might carelessly copy-paste this somewhere else without understanding the consequences.

But this code has _two_ problems: (1) it appears insecure, though it might actually be secure and (2) it doesn't pass the type checking.

I think that changing the relationship between JWTBearerSecurity from inheritance to dependency would be a minor change, and it would be manifestly correct.

Reply:
There is actually a third problem that it decodes the JWT twice, which is inefficient.

Reply:
My feeling here is that I fundamentally handled this wrong in my original pull request by having the JWT decode inside `JWTBearer`. `HTTPBearer` is supposed to pass through the Bearer JWT when `__call__` is executed. This is the root cause of the mypy issue, changing this behavior so we can pass decoded JWT to `ScopeAuthenticate`.

So this would suggest that the JWT decoding/signature checking should occur elsewhere, because yes, it doesn't make sense to have to decode a JWT twice. So it should happen in `ScopeAuthenticate`, which becomes a dependency for all authenticated endpoints. So the same code checking the scope is also checking the bearer token signature.

Reply:
See #1906.

Reply:
Yeah, like that.

Reply:
Yep, I reviewed the code last night but wanted to have a chance to run it, which I'm doing now...